File: backend/app/routers/agent.py
from typing import Any, Dict, Optional
from datetime import datetime
from bson import ObjectId

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from app.agents import build_agent_graph
from app.config import settings
from app.database import get_database
from app.models.memory import MemoryField


router = APIRouter(prefix="/agents", tags=["agents"])


class AgentRequest(BaseModel):
    message: str
    trip_id: Optional[str] = None  # Trip ID for auto-updating memory
    trip_context: Optional[Dict[str, Any]] = None
    trip_memory: Optional[Dict[str, Any]] = None


class AgentResponse(BaseModel):
    clarification: Optional[str]
    response: Optional[str]
    intent: Optional[Dict[str, Any]]
    task_plan: Optional[Dict[str, Any]]
    completed_tasks: Dict[str, Any]


@router.post("/plan", response_model=AgentResponse)
async def run_agent_workflow(payload: AgentRequest) -> AgentResponse:
    if not settings.openai_api_key:
        raise HTTPException(status_code=500, detail="OPENAI_API_KEY is not configured")

    graph = build_agent_graph()
    initial_state = {
        "user_message": payload.message,
        "trip_context": payload.trip_context or {},
        "trip_memory": payload.trip_memory or {},
        "intent": None,
        "task_plan": None,
        "clarification": None,
        "completed_tasks": {},
        "final_response": None,
    }

    final_state = await graph.ainvoke(initial_state)

    intent = final_state.get("intent")
    task_plan = final_state.get("task_plan")

    # Auto-update trip memory from intent if trip_id is provided
    if payload.trip_id and intent:
        # Use message text as a reference for sources (could be enhanced with actual message ID)
        await _update_memory_from_intent(payload.trip_id, intent, payload.message[:50])

    # Auto-generate plan from task execution if trip_id is provided and tasks completed
    completed_tasks = final_state.get("completed_tasks", {})
    if payload.trip_id and completed_tasks and intent:
        await _generate_plan_from_tasks(payload.trip_id, intent, completed_tasks)

    return AgentResponse(
        clarification=final_state.get("clarification"),
        response=final_state.get("final_response"),
        intent=intent.model_dump() if intent else None,
        task_plan=task_plan.model_dump() if task_plan else None,
        completed_tasks=final_state.get("completed_tasks", {}),
    )


async def _update_memory_from_intent(trip_id: str, intent: Any, message_ref: str):
    """Update trip memory from agent intent extraction"""
    try:
        db = get_database()
        memory_updates = {}
        
        # Update destination
        if intent.destinations and len(intent.destinations) > 0:
            destination_str = ", ".join(intent.destinations)
            memory_updates["destination"] = {
                "value": destination_str,
                "confidence": 80,  # Default confidence for new extractions
                "sources": [message_ref] if message_ref else []
            }
        
        # Update dates
        if intent.start_date or intent.end_date:
            dates_str = ""
            if intent.start_date and intent.end_date:
                dates_str = f"{intent.start_date} to {intent.end_date}"
            elif intent.start_date:
                dates_str = f"Starting {intent.start_date}"
            elif intent.end_date:
                dates_str = f"Ending {intent.end_date}"
            
            if dates_str:
                memory_updates["dates"] = {
                    "value": dates_str,
                    "confidence": 85,
                    "sources": [message_ref] if message_ref else []
                }
        
        # Update duration
        if intent.duration_days:
            memory_updates["duration"] = {
                "value": f"{intent.duration_days} days",
                "confidence": 80,
                "sources": [message_ref] if message_ref else []
            }
        
        # Update budget
        if intent.budget_total:
            memory_updates["budget"] = {
                "value": f"${intent.budget_total} total",
                "confidence": 75,
                "sources": [message_ref] if message_ref else []
            }
        elif intent.budget_per_person and intent.group_size:
            total = intent.budget_per_person * intent.group_size
            memory_updates["budget"] = {
                "value": f"${intent.budget_per_person} per person (${total} total)",
                "confidence": 75,
                "sources": [message_ref] if message_ref else []
            }
        
        # Update pace (from interests/constraints)
        if intent.interests:
            pace_hints = ["relaxed", "slow", "leisurely", "fast", "packed", "intensive"]
            pace_value = None
            for interest in intent.interests:
                interest_lower = interest.lower()
                if any(hint in interest_lower for hint in pace_hints):
                    pace_value = interest
                    break
            
            if not pace_value:
                # Default based on number of interests
                pace_value = "moderate" if len(intent.interests) <= 3 else "active"
            
            memory_updates["pace"] = {
                "value": pace_value,
                "confidence": 60,
                "sources": [message_ref] if message_ref else []
            }
        
        # Update memory in database if there are updates
        if memory_updates:
            memory_updates["updatedAt"] = datetime.utcnow()
            
            # Merge with existing memory, updating confidence and sources
            existing_memory = await db.trip_memory.find_one({"tripId": ObjectId(trip_id)})
            
            if existing_memory:
                # Merge updates, increasing confidence if value already exists
                for key, new_value in memory_updates.items():
                    if key == "updatedAt":
                        continue
                    existing_value = existing_memory.get(key)
                    if existing_value and existing_value.get("value") == new_value["value"]:
                        # Same value, increase confidence
                        new_value["confidence"] = min(100, existing_value.get("confidence", 0) + 10)
                        new_value["sources"] = list(set(existing_value.get("sources", []) + new_value["sources"]))
            
            await db.trip_memory.update_one(
                {"tripId": ObjectId(trip_id)},
                {"$set": memory_updates},
                upsert=True
            )
    except Exception as e:
        # Log error but don't fail the agent workflow
        print(f"Error updating memory from intent: {e}")


async def _generate_plan_from_tasks(trip_id: str, intent: Any, completed_tasks: Dict[str, Any]):
    """Generate a plan from agent task execution results"""
    try:
        db = get_database()
        
        # Only generate plan if itinerary tasks were completed
        has_itinerary_tasks = any(
            key.startswith("plan_day_") or key == "search_attractions"
            for key in completed_tasks.keys()
        )
        
        if not has_itinerary_tasks:
            return
        
        # Build itinerary structure from completed tasks
        itinerary: Dict[str, Any] = {}
        days = intent.duration_days or 1
        
        # Extract day plans from completed tasks
        for day in range(1, days + 1):
            day_key = f"plan_day_{day}"
            if day_key in completed_tasks:
                day_data = completed_tasks[day_key]
                if isinstance(day_data, dict) and "status" not in day_data:
                    itinerary[f"day_{day}"] = day_data
                else:
                    # Fallback: create basic day structure
                    itinerary[f"day_{day}"] = {
                        "title": f"Day {day}",
                        "activities": day_data.get("activities", []) if isinstance(day_data, dict) else [],
                        "cost": day_data.get("cost", "$0") if isinstance(day_data, dict) else "$0"
                    }
        
        # If no day plans, create a basic structure from attractions
        if not itinerary and "search_attractions" in completed_tasks:
            attractions_data = completed_tasks.get("search_attractions", {})
            if isinstance(attractions_data, dict):
                attractions = attractions_data.get("attractions", []) or []
                # Distribute attractions across days
                per_day = max(1, len(attractions) // days) if attractions else 0
                for day in range(1, days + 1):
                    start_idx = (day - 1) * per_day
                    day_attractions = attractions[start_idx:start_idx + per_day] if attractions else []
                    itinerary[f"day_{day}"] = {
                        "title": f"Day {day}",
                        "activities": [attr.get("name", str(attr)) if isinstance(attr, dict) else str(attr) for attr in day_attractions],
                        "cost": "$0"
                    }
        
        # Calculate budget breakdown if available
        budget = {
            "total": 0,
            "accommodation": 0,
            "activities": 0,
            "food": 0,
            "transport": 0
        }
        
        # Extract budget from tasks
        if "search_hotels" in completed_tasks:
            hotels_data = completed_tasks.get("search_hotels", {})
            if isinstance(hotels_data, dict):
                budget["accommodation"] = hotels_data.get("total_cost", 0) or 0
        
        # Calculate total from day costs
        total_cost = 0
        for day_data in itinerary.values():
            if isinstance(day_data, dict):
                cost_str = day_data.get("cost", "$0")
                cost = float(cost_str.replace("$", "").replace(",", "")) if isinstance(cost_str, str) else (cost_str if isinstance(cost_str, (int, float)) else 0)
                total_cost += cost
        
        budget["total"] = total_cost
        
        # Add budget to itinerary
        itinerary["budget"] = budget
        
        # Get latest version to increment
        latest = await db.plan_versions.find_one(
            {"tripId": ObjectId(trip_id)},
            sort=[("version", -1)]
        )
        next_version = (latest["version"] + 1) if latest else 1
        now = datetime.utcnow()
        
        # Create plan version
        plan_doc = {
            "tripId": ObjectId(trip_id),
            "version": next_version,
            "itinerary": itinerary,
            "createdBy": "agent",
            "createdAt": now,
        }
        
        await db.plan_versions.insert_one(plan_doc)
    except Exception as e:
        # Log error but don't fail the agent workflow
        print(f"Error generating plan from tasks: {e}")

File: backend/src/routers/agent.ts
import { Router, Response } from 'express';
import { ObjectId } from 'mongodb';
import { getDatabase } from '../database.js';
import { getCurrentUserId, AuthRequest } from '../utils/auth.js';
import { checkTripAccess } from '../utils/trip_permissions.js';
import { buildAgentGraph, TripIntent, ExecutionState } from '../agents/langgraph_workflow.js';
import { sseService } from '../services/sse.js';
import { z } from 'zod';

const router = Router();

const AgentRequestSchema = z.object({
  message: z.string(),
  trip_id: z.string().optional(),
  trip_context: z.record(z.any()).optional(),
  trip_memory: z.record(z.any()).optional(),
});

router.post('/plan', getCurrentUserId, async (req: AuthRequest, res: Response) => {
  try {
    if (!process.env.OPENAI_API_KEY) {
      res.status(500).json({ detail: 'OPENAI_API_KEY is not configured' });
      return;
    }

    const validated = AgentRequestSchema.parse(req.body);
    
    // If trip_id is provided, verify access
    if (validated.trip_id) {
      await checkTripAccess(validated.trip_id, req.userId!);
    }

    const graph = buildAgentGraph();
    const initialState: ExecutionState = {
      user_message: validated.message,
      trip_context: validated.trip_context || {},
      trip_memory: validated.trip_memory || {},
      completed_tasks: {},
    };

    const finalState = await graph.invoke(initialState);

    const intent = finalState.intent;
    const taskPlan = finalState.task_plan;

    // Auto-update trip memory from intent if trip_id is provided
    if (validated.trip_id && intent) {
      await updateMemoryFromIntent(validated.trip_id, intent, validated.message.substring(0, 50));
    }

    // Auto-generate plan from task execution if trip_id is provided and tasks completed
    const completedTasks = finalState.completed_tasks || {};
    let planVersionId: string | null = null;
    if (validated.trip_id && Object.keys(completedTasks).length > 0 && intent) {
      planVersionId = await generatePlanFromTasks(validated.trip_id, intent, completedTasks);
    }

    // Extract formatted flight data if available
    const formattedFlights = completedTasks.display_flights?.status === 'success' 
      ? completedTasks.display_flights.data 
      : null;
    
    if (formattedFlights) {
      console.log(`[AGENT_ROUTER] Found ${formattedFlights.length} formatted flight(s) for UI display`);
    } else {
      console.log('[AGENT_ROUTER] No formatted flights available (display_flights may not have been called or failed)');
    }

    res.json({
      clarification: finalState.clarification,
      response: finalState.final_response,
      plan_version_id: planVersionId,
      flights: formattedFlights, // Add structured flight data for UI
      intent: intent ? {
        original_message: intent.original_message,
        destinations: intent.destinations,
        origin: intent.origin,
        start_date: intent.start_date,
        end_date: intent.end_date,
        duration_days: intent.duration_days,
        group_size: intent.group_size,
        traveler_names: intent.traveler_names,
        budget_total: intent.budget_total,
        budget_per_person: intent.budget_per_person,
        interests: intent.interests,
        constraints: intent.constraints,
        requested_tasks: intent.requested_tasks,
        clarifications_needed: intent.clarifications_needed,
      } : null,
      task_plan: taskPlan ? {
        tasks: taskPlan.tasks,
        clarification_required: taskPlan.clarification_required,
        clarification_message: taskPlan.clarification_message,
      } : null,
      completed_tasks: completedTasks,
    });
  } catch (error: any) {
    if (error.name === 'ZodError') {
      res.status(400).json({ detail: 'Validation error', errors: error.errors });
      return;
    }
    if (error.message === 'Trip not found' || error.message === 'Access denied') {
      res.status(error.message === 'Trip not found' ? 404 : 403).json({ detail: error.message });
    } else {
      console.error('Agent workflow error:', error);
      res.status(500).json({ detail: error.message });
    }
  }
});

async function updateMemoryFromIntent(tripId: string, intent: TripIntent, messageRef: string): Promise<void> {
  try {
    const db = getDatabase();
    const memoryUpdates: Record<string, any> = {};

    // Update origin
    if (intent.origin) {
      memoryUpdates.origin = {
        value: intent.origin,
        confidence: 80,
        sources: messageRef ? [messageRef] : [],
      };
    }

    // Update destination
    if (intent.destinations && intent.destinations.length > 0) {
      const destinationStr = intent.destinations.join(', ');
      memoryUpdates.destination = {
        value: destinationStr,
        confidence: 80,
        sources: messageRef ? [messageRef] : [],
      };
    }

    // Update dates - store both formatted string and structured dates
    if (intent.start_date || intent.end_date) {
      let datesStr = '';
      if (intent.start_date && intent.end_date) {
        datesStr = `${intent.start_date} to ${intent.end_date}`;
      } else if (intent.start_date) {
        datesStr = `Starting ${intent.start_date}`;
      } else if (intent.end_date) {
        datesStr = `Ending ${intent.end_date}`;
      }

      if (datesStr) {
        memoryUpdates.dates = {
          value: datesStr,
          confidence: 85,
          sources: messageRef ? [messageRef] : [],
          // Also store structured dates for easier retrieval
          start_date: intent.start_date,
          end_date: intent.end_date,
        };
      }
    }

    // Update duration
    if (intent.duration_days) {
      memoryUpdates.duration = {
        value: `${intent.duration_days} days`,
        confidence: 80,
        sources: messageRef ? [messageRef] : [],
      };
    }

    // Update budget
    if (intent.budget_total) {
      memoryUpdates.budget = {
        value: `$${intent.budget_total} total`,
        confidence: 75,
        sources: messageRef ? [messageRef] : [],
      };
    } else if (intent.budget_per_person && intent.group_size) {
      const total = intent.budget_per_person * intent.group_size;
      memoryUpdates.budget = {
        value: `$${intent.budget_per_person} per person ($${total} total)`,
        confidence: 75,
        sources: messageRef ? [messageRef] : [],
      };
    }

    // Update pace (from interests/constraints)
    if (intent.interests && intent.interests.length > 0) {
      const paceHints = ['relaxed', 'slow', 'leisurely', 'fast', 'packed', 'intensive'];
      let paceValue: string | undefined;
      
      for (const interest of intent.interests) {
        const interestLower = interest.toLowerCase();
        if (paceHints.some(hint => interestLower.includes(hint))) {
          paceValue = interest;
          break;
        }
      }

      if (!paceValue) {
        paceValue = intent.interests.length <= 3 ? 'moderate' : 'active';
      }

      memoryUpdates.pace = {
        value: paceValue,
        confidence: 60,
        sources: messageRef ? [messageRef] : [],
      };
    }

    // Update memory in database if there are updates
    if (Object.keys(memoryUpdates).length > 0) {
      memoryUpdates.updatedAt = new Date();

      // Merge with existing memory, updating confidence and sources
      const existingMemory = await db.collection('trip_memory').findOne({ tripId: new ObjectId(tripId) });

      if (existingMemory) {
        // Merge updates, increasing confidence if value already exists
        for (const [key, newValue] of Object.entries(memoryUpdates)) {
          if (key === 'updatedAt') continue;
          
          const existingValue = existingMemory[key];
          if (existingValue && existingValue.value === newValue.value) {
            // Same value, increase confidence
            newValue.confidence = Math.min(100, (existingValue.confidence || 0) + 10);
            newValue.sources = [...new Set([...(existingValue.sources || []), ...newValue.sources])];
          }
        }
      }

      await db.collection('trip_memory').updateOne(
        { tripId: new ObjectId(tripId) },
        { $set: memoryUpdates },
        { upsert: true }
      );
    }
  } catch (error) {
    // Log error but don't fail the agent workflow
    console.error('Error updating memory from intent:', error);
  }
}

async function generatePlanFromTasks(
  tripId: string,
  intent: TripIntent,
  completedTasks: Record<string, any>
): Promise<string | null> {
  try {
    const db = getDatabase();

    // Only generate plan if itinerary tasks were completed
    const hasItineraryTasks = Object.keys(completedTasks).some(
      key => key.startsWith('plan_day_') || key === 'search_attractions'
    );

    if (!hasItineraryTasks) {
      return;
    }

    // Build itinerary structure from completed tasks
    const itinerary: Record<string, any> = {};
    const days = intent.duration_days || 1;

    // Extract day plans from completed tasks
    for (let day = 1; day <= days; day++) {
      const dayKey = `plan_day_${day}`;
      if (dayKey in completedTasks) {
        const dayData = completedTasks[dayKey];
        if (typeof dayData === 'object' && dayData !== null && !('status' in dayData)) {
          itinerary[`day_${day}`] = dayData;
        } else {
          // Fallback: create basic day structure
          itinerary[`day_${day}`] = {
            title: `Day ${day}`,
            activities: (typeof dayData === 'object' && dayData?.activities) ? dayData.activities : [],
            cost: (typeof dayData === 'object' && dayData?.cost) ? dayData.cost : '$0',
          };
        }
      }
    }

    // If no day plans, create a basic structure from attractions
    if (Object.keys(itinerary).length === 0 && 'search_attractions' in completedTasks) {
      const attractionsData = completedTasks.search_attractions;
      if (typeof attractionsData === 'object' && attractionsData !== null) {
        const attractions = attractionsData.attractions || [];
        // Distribute attractions across days
        const perDay = attractions.length > 0 ? Math.max(1, Math.floor(attractions.length / days)) : 0;
        for (let day = 1; day <= days; day++) {
          const startIdx = (day - 1) * perDay;
          const dayAttractions = attractions.slice(startIdx, startIdx + perDay);
          itinerary[`day_${day}`] = {
            title: `Day ${day}`,
            activities: dayAttractions.map((attr: any) =>
              typeof attr === 'object' ? attr.name || String(attr) : String(attr)
            ),
            cost: '$0',
          };
        }
      }
    }

    // Calculate budget breakdown if available
    const budget: Record<string, number> = {
      total: 0,
      accommodation: 0,
      activities: 0,
      food: 0,
      transport: 0,
    };

    // Extract budget from tasks
    if ('search_hotels' in completedTasks) {
      const hotelsData = completedTasks.search_hotels;
      if (typeof hotelsData === 'object' && hotelsData !== null) {
        budget.accommodation = hotelsData.total_cost || 0;
      }
    }

    // Calculate total from day costs
    let totalCost = 0;
    for (const dayData of Object.values(itinerary)) {
      if (typeof dayData === 'object' && dayData !== null) {
        const costStr = dayData.cost || '$0';
        const cost = typeof costStr === 'string'
          ? parseFloat(costStr.replace('$', '').replace(/,/g, '')) || 0
          : typeof costStr === 'number' ? costStr : 0;
        totalCost += cost;
      }
    }

    budget.total = totalCost;
    itinerary.budget = budget;

    // Get latest version to increment
    const latest = await db.collection('plan_versions')
      .findOne(
        { tripId: new ObjectId(tripId) },
        { sort: { version: -1 } }
      );

    const nextVersion = latest ? latest.version + 1 : 1;
    const now = new Date();

    // Create plan version
    const planDoc = {
      tripId: new ObjectId(tripId),
      version: nextVersion,
      itinerary,
      createdBy: 'agent',
      createdAt: now,
    };

    const result = await db.collection('plan_versions').insertOne(planDoc);
    
    // Broadcast plan update via SSE
    const planResponse = {
      id: result.insertedId.toString(),
      trip_id: tripId,
      version: planDoc.version,
      itinerary: planDoc.itinerary,
      created_by: planDoc.createdBy,
      created_at: now,
    };
    sseService.broadcastPlan(tripId, planResponse);
    
    return result.insertedId.toString();
  } catch (error) {
    // Log error but don't fail the agent workflow
    console.error('Error generating plan from tasks:', error);
    return null;
  }
}

export default router;
